### PR TITLE
add vim.ui.pick

### PR DESCRIPTION
**Pull request by teto**
I've finally snapped, every plugin has a different picker interface and so I dont know where to look: I would like all plugins to use the same selection mechanism.
It's a very rough draft (needs tests etc) but having a first version of `vim.ui.pick` would already improve the situation a lot, we can then refine it before the 0.6 release..

It defaults to inputlist for now but there is a project to add vim.ui components so we could have a floating window component that replaces it when loading runtime. Now I wonder if picking should be a UI event too.

right now in my init.lua I tested with:
```
if vim.ui then


vim.ui.pick = function (entries, opts)
	acceptable_files = vim.tbl_values(entries)


	local pickers = require "telescope.pickers"
	local finders = require "telescope.finders"
	local actions = require "telescope.actions"

	pickers.new({
	prompt_title = "test vim.pick",
	finder = finders.new_table {
	results = acceptable_files,
	entry_maker = function(line)
	  return {
		value = line,
	    ordinal = line,
	    display = line,
	  }
	end,

    attach_mappings = function(_)
      actions.select_default:replace(actions.run_builtin)
      return true
    end,
	}
}):find()
```
@tjdevries I actually spent more time writing the telescope picker than the neovim changes, and right now upon selection it opens the file but I just would like to return the selected entry, how can I do it ?

### CODE REVIEW

File: src/nvim/api/vim.c [hunk 1]
```diff
@@ -575,6 +575,25 @@ Object nvim_notify(String msg, Integer log_level, Dictionary opts, Error *err)
   return nlua_exec(STATIC_CSTR_AS_STRING("return vim.notify(...)"), args, err);
 }
 
+/// Notify the user with a message
+///
+/// Relays the call to vim.notify . By default forwards your message in the
+/// echo area but can be overridden to trigger desktop notifications.
```
Comment (bew) [inline src/nvim/api/vim.c:581]: Hmm the doc is wrong here, it's not a notification system but a picker system, a way to select one thing in a list

File: src/nvim/lua/vim.lua [hunk 2]
```diff
@@ -424,6 +424,14 @@ function vim.notify(msg, log_level, _opts)
   end
 end
 
+vim.ui = {
+  --- Picker
+  --- without a runtime, relies on inputlist()
+  --  see :help nvim_pick
+  pick = function (entries, opts)
```
Comment (bew) [inline src/nvim/lua/vim.lua:431]: For opts I'm thinking we can have a `multiselect` option? But i'm not sure how to impl with inputlist..

**Comment by fdschmidt93** [comment]
As a non-CS person, I'm no expert on async I'm afraid. I've actually tried doing a bit what I think you are intending to do.

As I understand 

```lua
vim.ui = {
  --- Picker
  --- without a runtime, relies on inputlist()
  --  see :help nvim_pick
  pick = function (entries, opts)
    return vim.fn.inputlist(vim.tbl_values(entries))
  end
}
```

and your example code above, you would like to return the `action_state.get_selected_entry` from the `picker:find()` call. There's two issues.

First, your `attach_mapping` should go out of the inner brackets of `finder` to `picker:new{}`. So far so easy :) 

Second, what I've tried is something along these lines.
```lua
local pick = function(entries, opts)
  acceptable_files = vim.tbl_values(entries)
  local value
  pickers.new({
    prompt_title = "test vim.pick",
    finder = finders.new_table {
      results = acceptable_files,
      entry_maker = function(line)
        return {
          value = line,
          ordinal = line,
          display = line,
        }
      end,
    },
    attach_mappings = function(prompt_bufnr)
      action_set.select:replace(function(prompt_bufnr)
        local entry = action_state.get_selected_entry()
        actions.close(prompt_bufnr)
        value = entry.value
      end)
      -- avoid centering on <CR>
      actions.center:replace(function() end)
      return true
    end,
  }):find()
  -- how to wait correctly here?
  return value
end
```
Though the function returns `value = nil` before the `action` is executed. So in my snippet for `value` to be set correctly, you'd essentially have to wait until e.g. `<CR>` is hit on the entry you'd want.

In any case, if I understand [this](https://github.com/nvim-telescope/telescope.nvim/blob/69adcbca876bbd12cb2d120e403bf4377c34a5a2/lua/telescope/actions/init.lua#L382) correctly, this is an open problem generally not yet solved :sweat_smile:

File: src/nvim/api/vim.c (hunk 1 above)
Comment (teto) [reply src/nvim/api/vim.c]: thanks for reminding me I am a messy committer :p

File: src/nvim/lua/vim.lua (hunk 2 above)
Comment (teto) [reply src/nvim/lua/vim.lua:431]: agreed. I see this as part of the next breaking change update of the feature. For now I just would like to merge a basic one with "prompt" in options.

**Comment by mfussenegger** [comment]
>  actually spent more time writing the telescope picker than the neovim changes, and right now upon selection it opens the file but I just would like to return the selected entry, how can I do it ?

Afaik it is not possible to come up with a telescope based implementation for a `pick` method that directly returns the selection.
Telescope or other plugins utilize float windows and need to react to new user inputs - so the nvim event loop needs to continue running - that isn't possible if the `pick` method blocks.

I think there are only two real options:

- Add a callback argument
- Wrap everything in coroutines, like it is done in https://github.com/vijaymarupudi/nvim-fzf


These are the two implementations that I tend to use:


```lua
function M.pick_one_sync(items, prompt, label_fn)
  local choices = {prompt}
  for i, item in ipairs(items) do
    table.insert(choices, string.format('%d: %s', i, label_fn(item)))
  end
  local choice = vim.fn.inputlist(choices)
  if choice < 1 or choice > #items then
    return nil
  end
  return items[choice]
end


function M.pick_one(items, prompt, label_fn, cb)
  local result = M.pick_one_sync(items, prompt, label_fn)
  cb(result)
end
```


`prompt` and `label_fn` could be moved into a `opts` argument and `label_fn` can default to `tostring`.
What I like about this interface is that `items` can be an arbitrary list of tables and you get the selected table back.
If `items` is a list of strings, it makes it difficult for the user of the API to map the selected item back to some richer data structure.



The telescope implementation would then look like this: (I think, haven't tested it)


```lua
local finders = require'telescope.finders'
local sorters = require'telescope.sorters'
local actions = require'telescope.actions'
local pickers = require'telescope.pickers'
vim.ui.pick_one = function(items, prompt, label_fn, cb)
  local opts = {}
  pickers.new(opts, {
    prompt_title = prompt,
    finder    = finders.new_table {
      results = items,
      entry_maker = function(entry)
        return {
          value = entry,
          display = label_fn(entry),
          ordinal = label_fn(entry),
        }
      end,
    },
    sorter = sorters.get_generic_fuzzy_sorter(),
    attach_mappings = function(prompt_bufnr)
      actions.select_default:replace(function()
        local selection = actions.get_selected_entry(prompt_bufnr)
        actions.close(prompt_bufnr)

        cb(selection.value)
      end)

      return true
    end,
  }):find()
end
```

File: src/nvim/lua/vim.lua (hunk 2 above)
Comment (tjdevries) [reply src/nvim/lua/vim.lua:431]: Then this should be `pick_one` IMO

**Comment by tjdevries** [comment]
As mentioned above (and I can write more about this later) there is no way to do this just in a normal function easily with telescope (although I could attempt to write something makes it possible maybe -- but I'm confident it will be unsuccessful). The reason I'm confident it will be unsuccesfull is because I need to run other neovim things in the mean time, and you can't just _randomly_ stop your lua state to do that.

In general, the callback approach would be easy to incorporate with telescope and other things that react to the UI. @mfussenegger example looks like it would work just fine.

The top down approach could work if we eventually import that `plenary.async` lib into core, because that can convert any function that has an argument with a callback -> async function.

**Comment by bfredl** [comment]
we want to support modal execution eventually (i e nested invocation of the neovim event loop for any purpose) https://github.com/neovim/neovim/pull/10842, though I don't want to promise any timeframe for this, so it would be best if this could work without it (i e the callback approach, as mentioned) in the meanwhile.

**Comment by mfussenegger** [comment]
Merged https://github.com/neovim/neovim/pull/15771

**Comment by teto** [comment]
thanks for following up, I get easily distracted :)

**Pull request closed without merging**